### ISSUE TITLE
NFS-Ganesha for CEPHFS very slow and often stale

### ISSUE THREAD

**Issue by borisbardak** [nfs-ganesha/nfs-ganesha#499]
We are testing exporting cephfs with nfs-ganesha but perfomance are very poor. NFS-ganesha server is located on VM with 10Gb ethernet, 8 cores and 12GB of RAM. Also, cluster is pretty big(156 OSD, 250 TB on SSD disks, 10 Gb ethernet with separeted cluster and public network)

I am running fio test comparing rbd with xfs, cephfs directly mounted to client(mounted to mds) and cephfs thru nfs-ganesha with these mount options -o v4.1 . NFS-ganesha results are so bad. Am I missing something? Why is this so poor in performace comparing to directly mounted cephfs?

My performance test is running fio from 3 clients, running 4k and 8k block with seq and rand read and writes.

Nfs-ganesha version rpms are

nfs-ganesha-ceph-2.8.2-2.el7.x86_64
nfs-ganesha-2.8.2-2.el7.x86_64

CEPH cluster is 14.2.

MY ganesha.conf is

NFS_CORE_PARAM
{
        Enable_NLM = false;

        Enable_RQUOTA = false;

        Protocols = 4;
}

NFSv4
{

        RecoveryBackend = rados_ng;

        Minor_Versions =  1,2;
}

MDCACHE {
        Dir_Chunk = 0;

        NParts = 1;
        Cache_Size = 1;
}

EXPORT
{
        Export_ID=100;

        Protocols = 4;

        Transports = TCP;

        Path = /;

        Pseudo = /cephfs/;

        Access_Type = RW;

        Attr_Expiration_Time = 0;

                 Delegations = R;

        Squash = no_root_squash;
        FSAL {
                Name = CEPH;

        }
}

CEPH
{
         Ceph_Conf = /etc/ceph/ceph.conf;

}

RADOS_KV
{
}

RADOS_URLS
{
}
LOG {
        Default_Log_Level = WARN;
        Components {
        ALL = DEBUG;
        }
        Facility {
               name = FILE;
               destination = "/var/log/ganesha/ganesha.log";
               enable = active;
        }
}

Also, nfs-ganesha is often stale. This is from ganesha.log

28/10/2019 12:02:28 : epoch 5db6c4ca : ganesha-server1 : ganesha.nfsd-23095[reaper] reaper_run :CLIENT ID :DEBUG :Now checking NFS4 clients for expiration
28/10/2019 12:02:38 : epoch 5db6c4ca : ganesha-server1 : ganesha.nfsd-23095[reaper] reaper_run :CLIENT ID :DEBUG :Now checking NFS4 clients for expiration
28/10/2019 12:02:48 : epoch 5db6c4ca : ganesha-server1 : ganesha.nfsd-23095[reaper] reaper_run :CLIENT ID :DEBUG :Now checking NFS4 clients for expiration
28/10/2019 12:02:58 : epoch 5db6c4ca : ganesha-server1 : ganesha.nfsd-23095[reaper] reaper_run :CLIENT ID :DEBUG :Now checking NFS4 clients for expiration
28/10/2019 12:03:08 : epoch 5db6c4ca : ganesha-server1 : ganesha.nfsd-23095[reaper] reaper_run :CLIENT ID :DEBUG :Now checking NFS4 clients for expiration
28/10/2019 12:03:18 : epoch 5db6c4ca : ganesha-server1 : ganesha.nfsd-23095[reaper] reaper_run :CLIENT ID :DEBUG :Now checking NFS4 clients for expiration
28/10/2019 12:03:28 : epoch 5db6c4ca : ganesha-server1 : ganesha.nfsd-23095[reaper] reaper_run :CLIENT ID :DEBUG :Now checking NFS4 clients for expiration

**Comment by dang** [comment]
I'm not sure what the problem is.  I don't use cephfs myself, but there have been reports that Ganesha is quite fast on it (and even a few reports that it was faster than native CephFS mounts, although that strikes me as odd).

Some things to check:  Try mounting cephfs on the Ganesha server, and see if it's slow there.  That would rule out Ganesha's connection to the cluster.  Also, try exporting a VFS mount from the Ganesha, and test the same fio test against that.  This should give you an idea of the performance available between the Ganesha server and the client.  If one of those shows odd results, then it may not be a Ganesha problem.

One potential issue is that a direct CephFS mount uses RADOS to talk directly to the OSDs for actual data I/O (metadata is always through the MDS).  However, all I/O over NFS must go through the Ganesha server.  That makes it's network connection a bottleneck for I/O, compared to multiple clients spreading their load among multiple OSDs.  To match performance, the Ganesha server may need a faster network connection than the clients have.  Since you didn't post any numbers, this may not apply; obviously if you're getting a few hundred k/s, then the network can't be the bottleneck.

Another possibility is I/O contention on the VM host that's hosting Ganesha's server.  If other VMs are also I/O heavy, then the Ganesha server may not be getting enough I/O cycles to actually fill it's network.

Basically, there's a lot of moving pieces here, and you're going to have to start testing them in isolation to find where the bottleneck is.

**Comment by borisbardak** [comment]
Ok, so we did make some tests.
Test1 - cephfs is mounted to the ganesha server. 
Fio results are
read: IOPS=28.3k, BW=148MiB/s (155MB/s)(4096MiB/27746msec)
write: IOPS=22.0k, BW=120MiB/s (126MB/s)(4096MiB/34212msec)

Same results I get from every other server that I mounted cephfs. I want to have these results with ganesha.

Test2 - Client machine have mounted nfs-ganesha that is connected to cephfs.
FSAL 
{
Name = CEPH;
}

Fio  results:
read: IOPS=3304, BW=17.2MiB/s (18.0MB/s)(4096MiB/238010msec)
write: IOPS=3291, BW=17.1MiB/s (17.0MB/s)(4096MiB/238949msec)

Test3 - Client machine have mounted nfs-ganesha that is connected to VFS. VFS is cephfs mounted on ganesha server.

FSAL {
                name = VFS;
        }

read: IOPS=5206, BW=27.1MiB/s (28.4MB/s)(4096MiB/151061msec)
write: IOPS=5191, BW=27.0MiB/s (28.4MB/s)(4096MiB/151487msec)

Test4 - Client machine have mounted nfs-ganesha that is connected to VFS. VFS is tmpfs.

read: IOPS=15.3k, BW=79.6MiB/s (83.4MB/s)(4096MiB/51475msec)
write: IOPS=15.3k, BW=79.6MiB/s (83.4MB/s)(4096MiB/51474msec)


I am pretty stuck here. :(

**Comment by dang** [comment]
@jtlayton Any ideas?

**Comment by jtlayton** [comment]
TBQH, FSAL_CEPH is just not currently built for speed.

As @dang points out, all the file I/O gets routed through ganesha before being distributed to the OSDs. That's a lot of extra copies, plus an extra hop for each NFS READ/WRITE op. Also, FSAL_CEPH is built on top of libcephfs and its performance is pretty awful compared to kcephfs.

In fact, it you have the cycles to do so, it might be interesting to see what sort of performance you get from ceph-fuse mounts, both with and without ganesha layered on top. My guess is that FSAL_VFS over ceph-fuse would probably give you somewhat similar performance here.

**Comment by rafalop** [comment]
@borisbardak can you paste your fio command/config? i am guessing you are using small block, random io based on iops and bw numbers. also are your cephfs mounts in testing using ceph-fuse or kernel?

one reason your ganesha numbers might be bad is because your ganesha conf is set to debug for all components. this is a huge overhead for ganesha for each io, which is exacerbated when you do many small ios.

i suggest trying test again with all = info for your ganesha logging, particularly if you are using small block io.

**Comment by borisbardak** [comment]
This is my fio file
# cat fio-test-1G-GANESHA.fio
[global]
size=1G
directory=/TEST_GA/
ioengine=libaio
iodepth=64

[rand-read-4k-1GB]
rw=randread
bs=4k


[rand-read-8k-1GB]
rw=randread
bs=8k

[seq-read-4k-1GB]
rw=read
bs=4k

[seq-read-8k-1GB]
rw=read
bs=8k

[rand-write-4k-1GB]
rw=randwrite
bs=4k


[rand-write-8k-1GB]
rw=randwrite
bs=8k

[seq-write-4k-1GB]
rw=write
bs=4k

[seq-write-8k-1GB]
rw=write
bs=8k

I am calling fio test with this

fio --randrepeat=1 --direct=1 --gtod_reduce=1 fio-test-1G-GANESHA.fio --group_reporting

Ganesha log is set to info.

**Comment by borisbardak** [comment]
So, I redone tests with change in block size.  Test is the same : rand read, rseq read, rand write and seq write with 4k, 8k, 1M and 4M blocks.

This is the results

**CEPHFS mounted through kernel**
_test fio-test-1G-CEPHFS-1-4k.fio_
   read: IOPS=22.4k, BW=87.5MiB/s (91.8MB/s)(2048MiB/23404msec)
  write: IOPS=16.8k, BW=65.6MiB/s (68.8MB/s)(2048MiB/31224msec)
   READ: bw=87.5MiB/s (91.8MB/s), 87.5MiB/s-87.5MiB/s (91.8MB/s-91.8MB/s), io=2048MiB (2147MB), run=23404-23404msec
  WRITE: bw=65.6MiB/s (68.8MB/s), 65.6MiB/s-65.6MiB/s (68.8MB/s-68.8MB/s), io=2048MiB (2147MB), run=31224-31224msec

_test fio-test-1G-CEPHFS-2-8k.fio_
   read: IOPS=18.3k, BW=143MiB/s (150MB/s)(2048MiB/14298msec)
  write: IOPS=16.7k, BW=131MiB/s (137MB/s)(2048MiB/15692msec)
   READ: bw=143MiB/s (150MB/s), 143MiB/s-143MiB/s (150MB/s-150MB/s), io=2048MiB (2147MB), run=14298-14298msec
  WRITE: bw=131MiB/s (137MB/s), 131MiB/s-131MiB/s (137MB/s-137MB/s), io=2048MiB (2147MB), run=15692-15692msec

_test fio-test-1G-CEPHFS-3-1M.fio_
   read: IOPS=623, BW=623MiB/s (653MB/s)(2048MiB/3287msec)
  write: IOPS=548, BW=549MiB/s (575MB/s)(2048MiB/3732msec)
   READ: bw=623MiB/s (653MB/s), 623MiB/s-623MiB/s (653MB/s-653MB/s), io=2048MiB (2147MB), run=3287-3287msec
  WRITE: bw=549MiB/s (575MB/s), 549MiB/s-549MiB/s (575MB/s-575MB/s), io=2048MiB (2147MB), run=3732-3732msec

_test fio-test-1G-CEPHFS-4-4M.fio_
   read: IOPS=159, BW=638MiB/s (669MB/s)(2048MiB/3212msec)
  write: IOPS=167, BW=670MiB/s (702MB/s)(2048MiB/3057msec)
   READ: bw=638MiB/s (669MB/s), 638MiB/s-638MiB/s (669MB/s-669MB/s), io=2048MiB (2147MB), run=3212-3212msec
  WRITE: bw=670MiB/s (702MB/s), 670MiB/s-670MiB/s (702MB/s-702MB/s), io=2048MiB (2147MB), run=3057-3057msec


**NFS-GANESHA mount**

_test fio-test-1G-GANESHA-1-4k.fio_
   read: IOPS=3379, BW=13.2MiB/s (13.8MB/s)(2048MiB/155136msec)
  write: IOPS=3378, BW=13.2MiB/s (13.8MB/s)(2048MiB/155174msec)
   READ: bw=13.2MiB/s (13.8MB/s), 13.2MiB/s-13.2MiB/s (13.8MB/s-13.8MB/s), io=2048MiB (2147MB), run=155136-155136msec
  WRITE: bw=13.2MiB/s (13.8MB/s), 13.2MiB/s-13.2MiB/s (13.8MB/s-13.8MB/s), io=2048MiB (2147MB), run=155174-155174msec


_test fio-test-1G-GANESHA-2-8k.fio_
   read: IOPS=3384, BW=26.4MiB/s (27.7MB/s)(2048MiB/77449msec)
  write: IOPS=3384, BW=26.4MiB/s (27.7MB/s)(2048MiB/77450msec)
   READ: bw=26.4MiB/s (27.7MB/s), 26.4MiB/s-26.4MiB/s (27.7MB/s-27.7MB/s), io=2048MiB (2147MB), run=77449-77449msec
  WRITE: bw=26.4MiB/s (27.7MB/s), 26.4MiB/s-26.4MiB/s (27.7MB/s-27.7MB/s), io=2048MiB (2147MB), run=77450-77450msec


_test fio-test-1G-GANESHA-3-1M.fio_
   read: IOPS=92, BW=92.9MiB/s (97.4MB/s)(2048MiB/22054msec)
  write: IOPS=91, BW=91.8MiB/s (96.3MB/s)(2048MiB/22303msec)
   READ: bw=92.9MiB/s (97.4MB/s), 92.9MiB/s-92.9MiB/s (97.4MB/s-97.4MB/s), io=2048MiB (2147MB), run=22054-22054msec
  WRITE: bw=91.8MiB/s (96.3MB/s), 91.8MiB/s-91.8MiB/s (96.3MB/s-96.3MB/s), io=2048MiB (2147MB), run=22303-22303msec

_test fio-test-1G-GANESHA-4-4M.fio_
   read: IOPS=28, BW=113MiB/s (118MB/s)(2048MiB/18164msec)
  write: IOPS=27, BW=112MiB/s (117MB/s)(2048MiB/18297msec)
   READ: bw=113MiB/s (118MB/s), 113MiB/s-113MiB/s (118MB/s-118MB/s), io=2048MiB (2147MB), run=18164-18164msec
  WRITE: bw=112MiB/s (117MB/s), 112MiB/s-112MiB/s (117MB/s-117MB/s), io=2048MiB (2147MB), run=18297-18297msec

As you can see from above results, ganesha is 5 to 8 times slower then kernel mounted CEPHFS. 

Ganesha server is even in public CEPH network with MTU 9000.

**Comment by ffilz** [comment]
Something to evaluate if we have resources to work on this.

**Comment by ffilz** [comment]
With some recent performance testing for FSAL_CEPH there may be energy to work on this.

**Comment by ffilz** [comment]
There is a newer issue regarding ganesha/cephfs performance #731. I'm going to close this one as a duplicate.
